Many things

CTalvio opened the pull request:
This list is off the top of my head. There is probably stuff I'm forgetting. I've done some testing and it all seems to work but there may be bugs.

**Disabled commands now remain accessible to admin**

**Map playlist order can now be random**

**Added !info and adapted lists into fully configurable info/welcome chat page**

**Vote kicking**
- Hidden from target
- Rejoin prevention (can persist for more than one match)
- Each player can only nominate one player for kicking at a time, no spamming
- Accepts partial names and ignores capitalization

**Votes**
- Set duration
- Set threshold

**Improved !mapvote**
- Block recent maps option
- No pagination on lists, always post all, colorcode
- Alias !nm and !maps
- Vote can be changed
- Display vote-pool info

**Vote RUI/Chat convar options**
- Use RUI/Chat UI or both

**Pagination**
- Ditched columns for row character limit
- Allow "all" switch to disable pagination and print full list
- Don't display page indicator for lists with only one page

**More color coding**
- Announce messages
- Errors
- Successes
- Enable hex color themeing using a convar
- Hex codes ignore capitalization

**ConVar tools**
- Can save into a relevant setting convar, no extra convars to confuse users
- Save array to convar
- Save array array to convar
- Pull array from convar
- Pull setting from convar
- Pull a valid map array from barely readable convars lol

CTalvio committed:
Change discord command to actually say "Discord:"

CTalvio committed:
More COLORS!!!

CTalvio committed:
Add convar toolset

Various functions for saving into and retrieving data from convars

CTalvio committed:
Typo

CTalvio committed:
Merge branch 'F1F7Y:main' into main

CTalvio committed:
Fix broken shit

CTalvio committed:
Welp, nevermind keeping this stuff organized

CTalvio committed:
Color codes, typos and wording

CTalvio committed:
So much stuff

Global chat messages are decoupled, for example if three votes are cast at the same time, only one chat message is sent to display the new count

RUI and/or CHAT messages for votes can be disabled and/or enabled

Votes now expire

Multiple kick votes are possible, but each player can only start one at a time

Pagination columns replaced with row length based division, page indicator only displayed when there is more than one present

Added "all" arg which can be used to ignore pagination and print a full list regardless of length

Maps order can now be randomized, recent maps can be blocked from coming up again within a number of matches

Private and Announced server messages now hae their own color coding

Error and success messages now have their own color codes

CTalvio committed:
Merge branch 'F1F7Y:main' into main

CTalvio committed:
Final touches

CTalvio committed:
Merge branch 'main' of https://github.com/CTalvio/ServerUtilities

CTalvio committed:
prevent empty final rows o pagination

F1F7Y commented on mods/Fifty.ServerUtilities/mod/scripts/vscripts/fsu_core.nut:
```diff
@@ -117,35 +121,87 @@ void function FSU_PrivateChatMessage( entity player, string message ) {
+	}
+	rowList.append(row)
+
+	return int( ceil( rowList.len() / ( rows * 1.0 ) ) ) // For some reason this does not use floats without the * 1.0 and as such would sometimes return zero when it shouldn't
```
You can use ` float( int )` to convert int to float. As `rows` and `rowList.len()` are ints the return type is int which is why it'd sometimes return 0.

CTalvio replied:
Will do. This issue was there before, it would cause there to be "0" pages if you tried to make them too dense and I tracked it down to this.

F1F7Y commented on mods/Fifty.ServerUtilities/mod/scripts/vscripts/fsu_core.nut:
```diff
@@ -52,12 +53,15 @@ void function FSU_PrintPrematchMessage() {
  * @param ownerHeaderCode The ANSI code to be used for the owner prefix ( "[OWNER]" )
 */
-void function FSU_SetTheme( string headerCode, string highlightCode, string textCode, string adminHeaderCode, string ownerHeaderCode ) {
+void function FSU_SetTheme( string headerCode, string highlightCode, string textCode, string adminHeaderCode, string ownerHeaderCode, string errorCode, string successCode, string announceCode ) {
```
I'll probably move this to ConVars as adding arguments to a function every update isn't backwards compatible.

F1F7Y commented on mods/Fifty.ServerUtilities/mod/scripts/vscripts/fsu_core.nut:
```diff
@@ -6,8 +6,9 @@ string highlight   = "\x1b[36m"
 string ownerHeader = "\x1b[92m"
-
-const FSU_LIST_ROWS = 5
+string error       = "\x1b[38;5;203m"
```
Not a big fan of these colours, as I said I'll move this to ConVars and probably make these fit together better.

CTalvio replied:
I can make the move to these being ConVars while I'm at it.
I could also set the new colors I added to just the same value as %T, making the colors by default be unchanged?

F1F7Y commented on mods/Fifty.ServerUtilities/mod/scripts/vscripts/fsu_core.nut:
```diff
@@ -95,9 +99,9 @@ void function FSU_ChatBroadcast( string message, bool usePopUp = false ) {
 	} else {
 		if( GetConVarBool( "FSU_PREFIX_SYSTEM_MESSAGES" ) )
-			Chat_ServerBroadcast( FSU_FormatString( "%F[FSU]%T " + message ), false )
+			Chat_ServerBroadcast( FSU_FormatString( "%F[FSU]%N " + message ), false )
```
![image](https://user-images.githubusercontent.com/64418963/214290459-0ea9bde8-2485-4c19-83f3-d2c85ac349f7.png)
Might be worth editing the convars to match this change.

CTalvio replied:
yes

CTalvio committed:
Add "commands" alias to !help

CTalvio committed:
Remove leftover code that does nothing

CTalvio committed:
Cleanup

CTalvio committed:
Do the thing

CTalvio committed:
Undid a war crime

Kate mixed indentation while editing, converted all my garbage to be the same as upstream

CTalvio committed:
Change examples to announce color

CTalvio committed:
Add a convar that can be used to set color theme

CTalvio committed:
Switch custom colors to use hex

Also non-capitalized codes work now

CTalvio committed:
Improve UI threads a bit

CTalvio committed:
Add a line instructing how to use !help, when you run !help

CTalvio committed:
Further improve pagination

CTalvio committed:
Fix some color codes and...

...make the maps list command color code recently played maps rather than display an extra list

CTalvio committed:
Make map chance percentage display only two digits

CTalvio committed:
Merge branch 'main' into main

CTalvio commented:
Anything else :D?

I have some more feature ideas but I'll leave that for another day.

CTalvio committed:
Merge branch 'main' into main

CTalvio committed:
Smol betterments

CTalvio committed:
Merge branch 'main' of https://github.com/CTalvio/ServerUtilities

CTalvio committed:
Adapt "list1-5" into completely configurable "info page" feature

CTalvio committed:
List other features that are not commands in readme

CTalvio committed:
Add toggle ConVar for !info

CTalvio committed:
Improve mapvote and map playlist

Map playlist convars and !nextmap now accept a far wider range of ways to write the name of a map

CTalvio committed:
Improve !kick name input

!kick now ignores capitalization and can find the correct player with just a part of a name (and throw an error if the arg applies to multiple players)

CTalvio commented:
Allrighty I think I'm for reals totally done finally completely. Probably.

CTalvio committed:
Be less spammy when user votes for an invalid map

CTalvio committed:
Do that last one even more better
